core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -284,12 +289,18 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
                   }
                   // If the broker is a follower, updates the isr in ZK and notifies the current leader
                   replicaStateMachine.handleStateChanges(Set(PartitionAndReplica(topicAndPartition.topic,
-                    topicAndPartition.partition, id)), OfflineReplica)
+                    topicAndPartition.partition, id)), OfflineReplica, newBatch = false)
                 }
               }
             }
           }
       }
+
+      inLock(controllerContext.controllerLock) {
```

Comment:
@lindong28 : So, you are saying that the CPU cost for checking the purgatory dominates the cost? The checkAndComplete() call to the purgatory is done per partition. So, if there are 1000 leader partitions whose leader need to be moved during controlled shutdown, the total number of checkAndComplete() calls seems to always be 1000 whether those partitions are included in 1000 LeaderAndIsrRequest or just 1 LeaderAndIsrRequest, right?